cpp/src/io/csv/datetime.cuh

```diff
@@ -49,100 +49,17 @@ __inline__ __device__ T to_non_negative_integer(char const* begin, char const* e
-  // In brief, the calculation is split over several components:
-  //     era: a 400 year range, where the date cycle repeats exactly
-  //     yoe: year within the 400 range of an era
-  //     doy: day within the 364 range of a year
-  //     doe: exact day within the whole era
-  // The months are shifted so that March is the starting month and February
-  // (possible leap day in it) is the last month for the linear calculation
-  year -= (month <= 2) ? 1 : 0;
-
-  const int32_t era = (year >= 0 ? year : year - 399_years) / 1_erasInYears;
-  const int32_t yoe = year - era * 1_erasInYears;
-  const int32_t doy = (153_days * (month + (month > 2 ? -3 : 9)) + 2) / 5 + day - 1;
-  const int32_t doe = (yoe * 365_days) + (yoe / 4_years) - (yoe / 100_years) + doy;
-
-  return (era * 1_erasInDays) + doe;
-}
-
-/**
- * @brief Computes the number of days since epoch, given a date.
- *
- * This function takes year, month, and day and returns the number of days since epoch (1970-01-01).
- *
- * @return days since epoch
- */
-__inline__ __device__ constexpr int32_t days_since_epoch(int year, int month, int day)
-{
-  // Shift the start date to epoch to match unix time
-  static_assert(static_cast<uint32_t>(days_since_baseline(1970, 1, 1)) == 719468_days,
-                "Baseline to epoch returns incorrect number of days");
-
-  return days_since_baseline(year, month, day) - days_since_baseline(1970, 1, 1);
-}
-
-/**
- * @brief Computes the number of seconds since epoch, given a date and time.
- *
- * This function takes year, month, day, hour, minute and second and returns
- * the number of seconds since epoch (1970-01-01),
- *
- * @return seconds since epoch
- */
-__inline__ __device__ constexpr int64_t seconds_since_epoch(
-  int year, int month, int day, int hour, int minute, int second)
-{
-  // Leverage the function to find the days since epoch
-  const int64_t days = days_since_epoch(year, month, day);
-
-  // Return sum total seconds from each time portion
-  return (days * 24 * 60 * 60) + (hour * 60 * 60) + (minute * 60) + second;
-}
-
 /**
  * @brief Extracts the Day, Month, and Year from a string.
  *
- * @param[in] begin Pointer to the first element of the string
- * @param[in] end Pointer to the first element after the string
- * @param[in] dayfirst Flag indicating that first field is the day
- * @param[out] year
- * @param[out] month
- * @param[out] day
- * @return true if successful, false otherwise
+ * @param begin Pointer to the first element of the string
+ * @param end Pointer to the first element after the string
+ * @param dayfirst Flag indicating that first field is the day
+ * @return Extracted year, month and day in `cuda::std::chrono::year_month_day` format
  */
-__inline__ __device__ bool extract_date(
-  char const* begin, char const* end, bool dayfirst, int* year, int* month, int* day)
+__inline__ __device__ cuda::std::chrono::year_month_day extract_date(char const* begin,
```

Comment:
@davidwendt is there common string to data parsing functionality that exists elsewhere that could be used here?

Reply:
The string to date conversion code is defined in https://github.com/rapidsai/cudf/blob/branch-21.12/cpp/src/strings/convert/convert_datetime.cu
The parsing here appears to be quite different.